### ISSUE TITLE
Debounce / sample stream combinators

### ISSUE THREAD

**Issue by NeoLegends** [alexcrichton/futures-rs#210]
Having a sample and debounce stream combinator would be really nice-to-have, especially in scenaria with user interaction. This would bring the library closer to other FRP-implementations.

Debounce works by ignoring values of a stream until a certain amount of time has passed without any new events. It is used by e.g. web developers when processing user input and ignoring the changes while the user is typing.
Sample works in a similar fashion: instead of ignoring all values the stream produces, it passes on values from the input stream at specified intervals. It can be used to ignore most intermediate values, yet provide the user with some feedback from time to time.

I don't know whether the suggestion is in the reach of this library since the implementations of both combinators are probably going to require a timer and an event loop, but this can be figured out, I guess.

**Comment by alexcrichton** [comment]
Sounds plausible to me! If it can't be implemented here then these may perhaps be best implementable at the tokio-core layer which has more access to scheduling and timers and such.

**Comment by NeoLegends** [comment]
Great! Since most FRP libraries feature a ton of combinators (_way_ more than futures-rs) and I'm sure not all of those combinators are applicable in Rust or don't really fit well into the zero-cost model, I suggest taking a look at http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html or https://msdn.microsoft.com/en-us/library/system.reactive.linq.observable(v=vs.103).aspx to get an idea of what is possible and which combinators could be ported.

**Comment by NeoLegends** [comment]
When the combinators to be ported are figured out, I could definetely start implementing some of them.

**Comment by alexcrichton** [comment]
Oh I at least personally don't mind implementing some more heavyweight combinators. Just because the core `Future` trait is zero-cost doesn't mean that everything around it needs to be as well! 

To that end I'd be down for basically implementing any useful combinator you've seen from other libraries, in my mind they're all welcome!

**Comment by alexcrichton** [comment]
Yeah if we want to implement these at the tokio-core layer we'd want an extension trait (or perhaps just free function constructors). I'd be fine exploring though to see if we can implement them at the futures layer as well. With `std::time::Instant` going all the way to tokio-core may not be necessary? Unsure though.

**Comment by NeoLegends** [comment]
I already thought about implementing the combinators only with `std::time` primitives, but the problem with that is that e.g. debounce needs to fire its last result _after_ the timeout has passed (asynchronously). This means we cannot rely on the underlying stream to notify the task of new items, which is why we're going to need a timer that `unpark()`s the task at the right time.

**Comment by siriux** [comment]
Wouldn't be a good idea to have Timeout and Interval from tokio-core directly in futures-rs? They can be useful for more than just IO, for example in GUIs.

**Comment by NeoLegends** [comment]
I guess the problem with that is that the timers themselves depend on a specific event loop, while futures-rs is agnostic to which one you use.

**Comment by siriux** [comment]
I see. Maybe it's a stupid idea but, could futures-rs include an abstraction that can be implemented on top of cpupool, tokio-core, ...? This way it would be usable with or without IO with the same interface.

**Comment by alexcrichton** [comment]
@siriux that's partly the topic of https://github.com/alexcrichton/futures-rs/issues/198 if we could add `park_timeout` generically (requiring event loops to implement it)

**Comment by carllerche** [comment]
There also is tokio-timer which is a free standing futures timer and is not bound to an event loop

**Comment by carllerche** [comment]
Personally, I wonder if there should be a futures-ext crate that can absorb additional dependencies and can act as a staging for combinators and other utilities as the landscape settles and it becomes more clear where everything goes

**Comment by NeoLegends** [comment]
Is there a rough ETA for the 0.2 release? https://github.com/tokio-rs/tokio-core/issues/69 has the unified timer story set for the 0.2 milestone, and we could just wait with the debounce / sample / buffer-for-time-combinators until that issue is resolved..?

**Comment by alexcrichton** [comment]
@NeoLegends unfortunately not yet, currently, but hopefully soon!

**Comment by NeoLegends** [comment]
Throttle has been merged into https://github.com/tokio-rs/tokio/commit/e166c4d91231c5af00df903b16b91c5c364f9b27.

**Comment by Nemo157** [comment]
One idea to decouple this from the underlying timers implementation: use an `impl Fn() -> impl Future<Output = ()>` to get the timer (alternatively an `impl Stream<Item = ()>` with the additional constraint that it will buffer at most a single item if not polled faster than its production interval). You could then provide the generic implementation in `futures`:

```rust
trait StreamExt {
    fn throttle(self, timer: impl Fn() -> impl Future<Output = ()>) -> impl Stream<Item = Self::Item> {
        ...
    }
}
```

(with some massaging of actual generic constraints since we can't yet use `impl Fn() -> impl Something`) and specific frameworks can simply wrap this with a function that uses their native timer functionality

```rust
trait tokio::StreamExt {
    fn throttle(self, duration: Duration) -> impl Stream<Item = Self::Item> {
        futures::StreamExt::throttle(self, move || Delay::new(clock::now() + duration))
    }
}
```

This would avoid duplicating the underlying functionality across multiple frameworks.

**Comment by NeoLegends** [comment]
Sounds like a valid point, though I'd change the signature of it a bit. Could you crosspost this to the relevant PR in the tokio repo?
